Alow setting the crossOrigin attribute for the image transform's image using a filter

fullofcaffeine opened the pull request:
## Description

~Introduce a new (optional) block editor property that can be used to specify the value of a `crossOrigin` attribute for the image element created by the image editor~. Allow setting the `crossOrigin` attribute for the image element used by the image transform feature using a WP JS filter. This allows it to be set when the app runs in a context where CORS is needed (i.e when the resulting image from the editing process ends up in a server with a domain that differs from the main blog domain) and decouples the logic that decides about the attribute / set it to the right value from Gutenberg.


## How has this been tested?

This should be useful in any WP environment that serves image assets from a different subdomain. 

### Reproducing the issue 

Here's how I reproduced the issue in a core WP instance (`5.7-alpha-49963`) using the Gutenberg build from this branch (I've used `wp-env`):

#### First of all, create two local domain mappings in your hosts file:

```
127.0.0.1 wp.local
127.0.0.1 uploads.wp.local
```

#### Go to `http://localhost:8888/wp-admin/options.php` and change the following settings to the following values:

* `upload_path` to `/var/www/html/wp-content/uploads`;
* `upload_url_path` to `http://uploads.wp.local:8888/wp-content/uploads`;

#### Edit your `wp-config.php` and change the following defines:

```
define( 'WP_SITEURL', 'http://wp.local:8888/' );
define( 'WP_HOME', 'http://wp.local:8888/' );
```

Now, if you try to add an image and edit to crop it, the browser will complain that it's not allowed because the CORS policy doesn't allow it.

### Trying the fix

#### Activate the hello test plugin and add the following filter to the end of `hello.php`:

```
function gutenberg_setup_image_editor_cors( $settings ) {
        $settings['imageEditingCrossOrigin'] = 'anonymous';
        return $settings;
};
add_filter( 'block_editor_settings', 'gutenberg_setup_image_editor_cors' );

function custom_headers() {
        header( 'Access-Control-Allow-Origin: http://wp.local:8888' );
        header( 'Vary: Origin' );
        header( 'Access-Control-Allow-Credentials: true' );
}

// send_headers doesn't work in wp-admin for some reason...
add_action( 'init', 'custom_headers' );
```

I'm sure there's a better place to add actions and filters that don't require a plugin, if so, let me know!

#### Test (currently failing, see [why](https://github.com/WordPress/gutenberg/pull/28255#issuecomment-763734258))

This should allow the image editing to work if the image is served from a different domain/subdomain. However, it still fails. The problem it seems is that the request that fetches the image** (see below) doesn't include the CORS headers set above. I don't know how to set the headers in WordPress so that it applies to the request that original fetches the image:

![foo](https://user-images.githubusercontent.com/81248/105119530-222e8380-5a96-11eb-9342-a11e0b9ae226.png)

As you can see, the actual post.php request does include the headers we've set above:

![foo1](https://user-images.githubusercontent.com/81248/105119592-45593300-5a96-11eb-8200-fecb9319f462.png).


## Types of changes

New feature (non-breaking change which adds functionality) 


## Checklist:
- [X] My code is tested.
- [X] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

fullofcaffeine committed:
Set the crossOrigin attribute for the image editing to the corresponding block setting value, if available

sirreal commented on packages/block-library/src/image/image-editing/use-transform-image.js:
```diff
@@ -12,6 +18,11 @@ function useTransformState( { url, naturalWidth, naturalHeight } ) {
 	const [ defaultAspect, setDefaultAspect ] = useState();
 
+	const { imageEditingCrossOriginValue } = useSelect( ( select ) => {
+		const { getSettings } = select( 'core/block-editor' );
+		return pick( getSettings(), [ 'imageEditingCrossOriginValue' ] );
```
I don't think we need `pick` here, checking for a `string` is probably sufficient. [The actual allowed values are only a few](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/crossOrigin).

```suggestion
		const value = select( 'core/block-editor' ).getSettings().imageEditingCrossOrigin;
		return typeof value === 'string' ? value : undefined;
```

fullofcaffeine replied:
Yeah, I had that on my list, I noticed `pick` would probably not be needed for a single value, though I was focused on making it work for the very first iteration. This change looks good. Thanks for the nudge!

sirreal commented on packages/block-library/src/image/image-editing/use-transform-image.js:
```diff
@@ -90,6 +101,9 @@ function useTransformState( { url, naturalWidth, naturalHeight } ) {
 		const el = new window.Image();
 		el.src = url;
 		el.onload = editImage;
+		if ( imageEditingCrossOriginValue ) {
```
We can't rely on a falsy check, `""` empty string is a valid value.

```suggestion
		if ( typeof imageEditingCrossOriginValue === 'string' ) {
```

fullofcaffeine replied:
👍🏻

sirreal commented on packages/editor/src/components/provider/use-block-editor-settings.js:
```diff
@@ -171,6 +171,7 @@ function useBlockEditorSettings( settings, hasTemplate ) {
 				'hasFixedToolbar',
 				'hasReducedUI',
 				'imageEditing',
+				'imageEditingCrossOriginValue',
```
```suggestion
				'imageEditingCrossOrigin',
```

fullofcaffeine replied:
👍🏻

sirreal reviewed (commented):
Thanks for starting on this. I left some early feedback.

fullofcaffeine committed:
Improve property name, value suffix is redundant

Co-authored-by: Jon Surrell <jon.surrell@automattic.com>

fullofcaffeine committed:
Avoid using `pick` and access the object directly and guard against undefined

Co-authored-by: Jon Surrell <jon.surrell@automattic.com>

fullofcaffeine committed:
Check if the value type of a string instead of relying on checking against falsey values

One of the valid values for `crossOrigin` is an empty string (`""`).

Co-authored-by: Jon Surrell <jon.surrell@automattic.com>

fullofcaffeine committed:
Get rid of the value suffix, it is redundant

Co-authored-by: Jon Surrell <jon.surrell@automattic.com>

fullofcaffeine committed:
Remove import for lodash's `pick`

Co-authored-by: Jon Surrell <jon.surrell@automattic.com>

fullofcaffeine committed:
Remove the value suffix for other occurrences of imageEditingCrossOriginValue

fullofcaffeine committed:
Do not destruct as select useSelect is returning the value directly and not the object

fullofcaffeine committed:
Fix property name in the block-library package README

noahtallen commented on packages/block-editor/src/store/defaults.js:
```diff
@@ -14,6 +14,7 @@ export const PREFERENCES_DEFAULTS = {
  * @property {boolean} alignWide Enable/Disable Wide/Full Alignments
  * @property {Array} availableLegacyWidgets Array of objects representing the legacy widgets available.
  * @property {boolean} imageEditing Image Editing settings set to false to disable.
+ * @property {string} [imageEditingCrossOrigin] Optional value to add to the image editing img el's `crossOrigin` attribute
```
Do we need to specify that this is specifically for image editing? I wonder if it would be better to set this more generally, like if this cross origin value would be used anywhere other than image editing... 🤔 

Other than that, this looks pretty great. I like that we are using settings for it, so that it's easy to control from the back end, and even the front end if we needed.

sirreal replied:
Yes, we should definitely go as far as `imageCrossOrigin`. I've been wondering how specific or generic to make this. Conceptually, I'd expect this to mean "will media be served from a CDN?" and it's something that the plugin or Core should be able to intelligently set.

fullofcaffeine replied:
Sounds good, that makes sense. I'll change the name 👍🏻

sirreal committed:
update docs

ockham commented:
> This should allow the image editing to work if the image is served from a different domain/subdomain. However, it still fails. The problem it seems is that the request that fetches the image** (see below) doesn't include the CORS headers set above. I don't know how to set the headers in WordPress so that it applies to the request that original fetches the image: [...]

I think that behavior is expected. As @ellatrix pointed out to me in a video call, image files are served straight by the server, with no involvement from WordPress. So in order to change their headers, we'd need to edit the server configuration (which is of cource outside the scope of Gutenberg and even WordPress).

fullofcaffeine committed:
Rename block attr to `imageCrossOrigin`, as it might be used beyond the image editor in the future

ellatrix commented on packages/block-library/src/image/image-editing/use-transform-image.js:
```diff
@@ -90,6 +97,9 @@ function useTransformState( { url, naturalWidth, naturalHeight } ) {
 		el.src = url;
 		el.onload = editImage;
+		if ( typeof imageCrossOrigin === 'string' ) {
+			el.crossOrigin = imageCrossOrigin;
```
I guess there's no one size fits all value for this? :)

WunderBart replied:
Actually, what if we set fixed `use-credentials` here? Is there a scenario where it wouldn't work like that?

sirreal replied:
I don't believe so. The server would need additional configuration to set [`Access-Control-Allow-Credentials`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Credentials). I don't think that's very common for image requests.

It's really something sites need to be able to configure.

fullofcaffeine replied:
Not sure I understood. Could you elaborate?

sirreal replied:
The question is whether we can just use a single value that works for everyone. I don't think that will work for all cases, and at the very least could have the undesirable side-effect of sending cookies when you didn't expect or need to.

ellatrix commented on packages/block-editor/README.md:
```diff
@@ -484,6 +484,7 @@ _Properties_
 -   _alignWide_ `boolean`: Enable/Disable Wide/Full Alignments
 -   _availableLegacyWidgets_ `Array`: Array of objects representing the legacy widgets available.
 -   _imageEditing_ `boolean`: Image Editing settings set to false to disable.
+-   _imageCrossOrigin_ `[string]`: Optional value for requesting images with CORS, allowing them to be used in a `<canvas>` as if they were loaded from the current origin.
```
Might be worth explaining what possible values there are.

ellatrix replied:
Oh, and is this restricted for use in images or can this also be used for other media in the future?

sirreal replied:
Video elements seem to use the same attribute, I'd say it's potentially any media that can be used in a canvas:

> `crossorigin`
> This enumerated attribute indicates whether to use CORS to fetch the related video. CORS-enabled resources can be reused in the <canvas> element without being tainted. The allowed values are:

Maybe `mediaCrossOrigin` is best. We'll work on better description.

sirreal replied:
Let's say the value should be one of:

- omitted (`undefined`)
- `""` or `"anonymous"`
- `"use-credentials"`

ellatrix commented on packages/block-library/src/image/image-editing/use-transform-image.js:
```diff
@@ -12,6 +13,12 @@ function useTransformState( { url, naturalWidth, naturalHeight } ) {
 	const [ defaultAspect, setDefaultAspect ] = useState();
 
+	const imageCrossOrigin = useSelect( ( select ) => {
+		const value = select( 'core/block-editor' ).getSettings()
```
Could we maybe use a filter in JS here, instead of using block settings?

sirreal replied:
That certainly seems technically possible. Do you think that would be better?

This seems to me like site configuration that the backend is able to make decisions about. It doesn't seem like the kind of thing the client side needs to be deciding at runtime.

fullofcaffeine replied:
I'm not familiar with WP js filters, would it be possible to pick up a value set from the server-side / PHP with it? Could you provide an example? Thanks!

ockham replied:
Filter seems to make sense to me, as it's more aligned with WordPress's typical way of extending/customizing things, whereas `block-settings` could be deemed an internal carrier of information (and thus, and implementation detail that might change).

On JS vs server-side/PHP filter: Agree that 

> It doesn't seem like the kind of thing the client side needs to be deciding at runtime.

A point could be made however that making it a JS filter makes some sense, as it "cuts out the middle man": It's a customization that needs to happen on the client side, so it's arguable that it doesn't have to be Gutenberg's responsibility to take on the chore of communicating it from the server to the client. (As long as that can be done easily enough by whatever plugin/extensions customizes this.)

(I don't feel very strongly about the JS vs PHP filter though.)

sirreal replied:
I've been reflecting on this and there _may_ be a good reason to handle it in JavaScript.

The case we're considering is when we're on our site (example.com) and we want to use images from our CDN on a different domain (example.net). That's easy to cover in this case because we understand usage at this time.

However, we only want to set `crossOrigin` requests for our known domains and not for arbitrary ones. We might want to `use-credentials` for our known `example.net`, but this hook could be used with other domains where we wouldn't want to send our credentials. We'd need to make this decision at runtime, or have some more complex data to describe crossOrigin based on domain:

```js
addFilter( 'imageTransform.crossOrigin', 'my-filter', ( crossOrigin, imgSrc ) => {
  if ( imgSrc.startsWith( 'https://example.net' ) ) {
    return 'use-credentials';
  }
  return crossOrigin;
} )
```

To acheive something similar via block editor settings we'd need something like the following:

```php
function configure_block_settings_media_crossorigin( $settings ) {
	$settings['mediaCrossOriginDomains'] = array(
		'example.net' => 'use-credentials',
	);
	return $settings;
};
add_filter( 'block_editor_settings', 'configure_block_settings_media_crossorigin' );
```

fullofcaffeine replied:
I've changed this to use a filter, like you suggested above, @sirreal. There's a nice side-effect in that it requires much less code :).

>  but this hook could be used with other domains where we wouldn't want to send our credentials. We'd need to make this decision at runtime, or have some more complex data to describe crossOrigin based on domain:

I'm not sure I fully understood what you meant here though. In the filter snippet above you're returning `use-credentials` if the `img` `src` matches a certain domain, and then `crossOrigin` as is, if not. At first sight, it didn't seem to match the logic we had before, so I instead tested with a filter that looks like this:

```
	addFilter(
		'gutenberg.imageTransform.crossOrigin',
		'gutenberg/imageTransform',
		( imgSrc ) => {
			if (
				// Only set crossOrigin for images served by us
				imgSrc.match( /https:\/\/.*\.files.adomain.com\/.*/ )
			) {
				return attrs.image_transform_cors_image_cross_origin; // this is calculated in the PHP backend and passed here using `wp_localize_script`. It returns `anonymous` or `use-credentials`.
			}
		}
	);
```

Let me know if this is what you had in mind. Thanks!

sirreal commented on packages/block-editor/src/store/defaults.js:
```diff
@@ -14,6 +14,7 @@ export const PREFERENCES_DEFAULTS = {
  * @property {boolean} alignWide Enable/Disable Wide/Full Alignments
  * @property {Array} availableLegacyWidgets Array of objects representing the legacy widgets available.
  * @property {boolean} imageEditing Image Editing settings set to false to disable.
+ * @property {string} [imageCrossOrigin] Optional value for requesting images with CORS, allowing them to be used in a `<canvas>` as if they were loaded from the current origin.
```
I believe this is the type we want:

```suggestion
 * @property { '' | 'anonymous' | 'use-credentials' | undefined } [imageCrossOrigin] Optional value for requesting images with CORS, allowing them to be used in a `<canvas>` as if they were loaded from the current origin.
```

sirreal commented on packages/block-library/src/image/image-editing/use-transform-image.js:
```diff
@@ -12,6 +13,12 @@ function useTransformState( { url, naturalWidth, naturalHeight } ) {
+	const imageCrossOrigin = useSelect( ( select ) => {
+		const value = select( 'core/block-editor' ).getSettings()
+			.imageCrossOrigin;
+		return typeof value === 'string' ? value : undefined;
```
I wonder if we should avoid getting anything other than allowed values 🤔 

The [spec](https://html.spec.whatwg.org/multipage/urls-and-fetching.html#cors-settings-attribute) describes valid values.

fullofcaffeine replied:
Good point.

fullofcaffeine committed:
Rename block setting to `mediaCrossOrigin` and list possible values for it in the jsdoc

Renamed because it might be used for other kind of assets in the future (i.e videos).

fullofcaffeine committed:
Only return a string value if it is part of the set of allowed values

sirreal commented on packages/block-library/src/image/image-editing/use-transform-image.js:
```diff
@@ -13,10 +13,13 @@ function useTransformState( { url, naturalWidth, naturalHeight } ) {
-		return typeof value === 'string' ? value : undefined;
+			.mediaCrossOrigin;
+		return typeof value === 'string' &&
+			value.includes( '', 'anonymous', 'use-credentials' )
```
This looks a bit backwards 🙂 
```suggestion
			[ '', 'anonymous', 'use-credentials' ].includes( value )
```

sirreal committed:
Fix includes check

WunderBart committed:
Sort the settings list

fullofcaffeine committed:
Use a WP JS filter to get the value of crossOrigin, if any

fullofcaffeine committed:
Pass undefined to the default crossOrigin value in the filter callback.

In this case, the logic to to decide the crossOrigin is in the filter function itself.

fullofcaffeine commented on packages/block-library/src/image/image-editing/use-transform-image.js:
```diff
@@ -90,6 +91,15 @@ function useTransformState( { url, naturalWidth, naturalHeight } ) {
+			undefined,
+			url
+		);
+		if ( typeof imgCrossOrigin === 'string' ) {
```
I think we could avoid this if by setting the return value of the filter directly to `crossOrigin` 🤔

ellatrix commented on packages/block-library/src/image/image-editing/use-transform-image.js:
```diff
@@ -90,6 +91,15 @@ function useTransformState( { url, naturalWidth, naturalHeight } ) {
 		el.onload = editImage;
+
+		const imgCrossOrigin = applyFilters(
+			'gutenberg.imageTransform.crossOrigin',
```
Can this be "media.crossOrigin" or something like that? How will we make sure the "gutenberg" part is updated or removed when this lands in core? Is there a standard way to name filters?

ellatrix replied:
Is there a place for documentation of filters?

sirreal replied:
Let's use `media.crossOrigin` and document it here https://github.com/WordPress/gutenberg/tree/004a32ec09d5b789be694d85725fe49cca385f08/docs/designers-developers/developers/filters 👍

ellatrix commented:
I have no objection for a filter. The only thing maybe missing here is some documentation.

fullofcaffeine committed:
Change filter namespace to `media.crossOrigin`

fullofcaffeine committed:
Add documentation for the media.crossOrigin filter

fullofcaffeine committed:
Update changelog

sirreal commented on docs/designers-developers/developers/filters/block-filters.md:
```diff
@@ -319,6 +319,30 @@ wp.hooks.addFilter( 'editor.BlockListBlock', 'my-plugin/with-client-id-class-nam
+	// the url to actual foreign media, useful if you want to decide
+	// the value of crossOrigin based upon it.
+	( crossOrigin, mediaSrc ) => {
+		if (mediaSrc.startsWith('https://my.cdn')) {
```
Let's always use safe example domains:
```suggestion
		if ( mediaSrc.startsWith( 'https://example.com' ) ) {
```

fullofcaffeine replied:
Oh, that's a good point. Thanks for the nudge.

sirreal commented on packages/block-library/CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@
 
+### New Features
+
+- Image block: Allow setting the `crossOrigin` ([#](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/crossOrigin), [#](https://developer.mozilla.org/en-US/docs/Web/HTML/CORS_enabled_image)) attribute for the image editor's `<img>` element using the `media.crossOrigin` filter (to allow it to work with foreign-origin images).
```
```suggestion
- Allow setting the `crossOrigin` attribute so the `useTransformImage` hook can use cross-origin sources ([#28255](https://github.com/WordPress/gutenberg/pull/28255/)).
```

sirreal reviewed (approved):
Two small suggestions for the docs, but this is tested and working well. Thank you!

sirreal commented on docs/designers-developers/developers/filters/block-filters.md:
```diff
@@ -319,6 +319,30 @@ wp.hooks.addFilter( 'editor.BlockListBlock', 'my-plugin/with-client-id-class-nam
+
+Used to set or modify the `crossOrigin` attribute for foreign-origin media elements (i.e  `<img>`, `<audio>` , `<img>` , `<link>` , `<script>`, `<video>`). See this [article](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/crossorigin) for more information the `crossOrigin` attribute, its values and how it applies to each element.
+
+One example of it in action is in the image block's transform feature, to allow foreign-loaded images to be used in a `<canvas>`.
```
```suggestion
One example of it in action is in the Image block's transform feature, to allow foreign-loaded images to be used in a `<canvas>`.
```

fullofcaffeine committed:
Improve and simplify the changelog

Co-authored-by: Jon Surrell <jon.surrell@automattic.com>

fullofcaffeine committed:
Add a few grammar improvements to the docs

Co-authored-by: Jon Surrell <jon.surrell@automattic.com>

fullofcaffeine committed:
Change example to use a safe `example.com` domain

Co-authored-by: Jon Surrell <jon.surrell@automattic.com>

sirreal commented:
This is ready to land, the failing tests seem to be failing on the main branch as well. I've rebased and will review test results to look for differences from the main branch failures then merge.

youknowriad commented:
There are three tests failing right now due to a Core regression in https://core.trac.wordpress.org/ticket/50025

inline-tokens, image size and classic block.

If these are the only failures right now, we can force merge.
